File: Level-2/selection_sort.py
#!/usr/bin/python

# Date: 2018-09-08
#
# Description:
# Implement selection sort.
#
# Approach:
# Finds index of minimum(when sorting in ascending) element from the remaining
# array and swap it with element at the end of sorted sub-array.
# While sorting in ascending order minimum element reaches at first place after
# first iteration of outer loop.
#
# Complexity:
# O(n^2)


def selection_sort_ascending(A):
    for i in range(len(A)):
        min_idx = i
        for j in range(i + 1, len(A)):
            if A[min_idx] > A[j]:
                min_idx = j
        A[i], A[min_idx] = A[min_idx], A[i]  # Swap min with sorted sub-array.

def selection_sort_descending(A):
    for i in range(len(A)):
        max_idx = i
        for j in range(i + 1, len(A)):
            if A[max_idx] < A[j]:
                max_idx = j
        A[max_idx], A[i] = A[i], A[max_idx]

def sort(A):
    B = A[:]
    selection_sort_ascending(A)
    selection_sort_descending(B)
    return (A, B)

assert sort([3, 4, 5, 2, 1]) == ([1, 2, 3, 4, 5], [5, 4, 3, 2, 1])
assert sort([3, 4, 5, 2, 1, 6]) == ([1, 2, 3, 4, 5, 6], [6, 5, 4, 3, 2, 1])
assert sort([]) == ([], [])
assert sort([1]) == ([1], [1])
assert sort([2, 1]) == ([1, 2], [2, 1])

File: Level-2/selection-sort.go
/*
Date: 2023-12-29

Description:
Implement selection sort.

Approach:
Finds index of minimum (when sorting in ascending) element from the remaining
array and swap it with element at the end of sorted sub-array.
While sorting in ascending order minimum element reaches at first place after
first iteration of outer loop.

Complexity:
O(n^2)
*/

package main

import (
	"fmt"
	"reflect"
)

// Sorts slice in increasing order using selection sort.
func selectionSort(nums []int) {
	for i := 0; i < len(nums); i++ {
		minIdx := i
		for j := i + 1; j < len(nums); j++ {
			// For sorting in descending order, find maxIdx.
			if nums[j] < nums[minIdx] {
				minIdx = j
			}
		}
		nums[minIdx], nums[i] = nums[i], nums[minIdx]
	}
}

func main() {
	testCases := []struct {
		input, expected []int
	}{
		{input: []int{5, 4, 3, 2, 1, 10, 28, 7, 6}, expected: []int{1, 2, 3, 4, 5, 6, 7, 10, 28}},
		{input: []int{3, 4, 5, 2, 1}, expected: []int{1, 2, 3, 4, 5}},
		{input: []int{3, 4, 5, 2, 1, 6}, expected: []int{1, 2, 3, 4, 5, 6}},
		{input: []int{}, expected: []int{}},
		{input: []int{1}, expected: []int{1}},
		{input: []int{2, 1}, expected: []int{1, 2}},
	}
	for idx, tc := range testCases {
		inputCopy := make([]int, len(tc.input))
		copy(inputCopy, tc.input)
		selectionSort(inputCopy)
		if !reflect.DeepEqual(inputCopy, tc.expected) {
			fmt.Printf("\nTC %d failed. Expected %v, got %v", idx, tc.expected, inputCopy)
		}
	}
}
